core/src/main/scala/kafka/utils/PasswordEncoder.scala

```diff
@@ -0,0 +1,160 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+package kafka.utils
+
+import java.nio.charset.StandardCharsets
+import java.security.{AlgorithmParameters, NoSuchAlgorithmException, SecureRandom}
+import java.security.spec.AlgorithmParameterSpec
+import javax.crypto.{Cipher, SecretKeyFactory}
+import javax.crypto.spec._
+
+import kafka.utils.PasswordEncoder._
+import org.apache.kafka.common.config.ConfigException
+import org.apache.kafka.common.config.types.Password
+import org.apache.kafka.common.utils.Base64
+
+import scala.collection.Map
+
+object PasswordEncoder {
+  val KeyFactoryAlgorithmProp = "keyFactoryAlgorithm"
+  val CipherAlgorithmProp = "cipherAlgorithm"
+  val InitializationVectorProp = "initializationVector"
+  val KeyLengthProp = "keyLength"
+  val SaltProp = "salt"
+  val IterationsProp = "iterations"
+  val EncyrptedPasswordProp = "encryptedPassword"
+  val PasswordLengthProp = "passwordLength"
+}
+
+class PasswordEncoder(secret: Password,
```

Comment:
I am wondering if this should be implemented in the common package in case we ever have need to reuse on the client side.

Reply:
At the moment, it is tailored for password encryption rather than generic encryption. If we need this functionality on the client-side, we can decide whether to reuse this or not. For now, shall I just leave it in core?